src/coreclr/jit/lowerxarch.cpp

```diff
@@ -3693,6 +3720,75 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
         LowerNode(cast);
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerAndOpToResetLowestSetBit: Lowers a tree AND(X, ADD(X, -1) to HWIntrinsic::ResetLowestSetBit
+//
+// Arguments:
+//    andNode - GentreePtr of binOp
+//
+// Return Value:
+//    Returns the replacement node if one is created else nullptr indicating no replacement
+//
+// Assumptions:
+//    andNode is GT_AND and (TYP_INT or TYP_LONG)
+//
+GenTree* Lowering::LowerAndOpToResetLowestSetBit(GenTree* andNode)
+{
+    assert(andNode->OperIs(GT_AND) && varTypeIsIntegral(andNode));
+
+    GenTree* op1 = andNode->gtGetOp1();
+    GenTree* op2 = andNode->gtGetOp2();
+    if (op1->OperIs(GT_LCL_VAR) && op2->OperIs(GT_ADD) && !comp->lvaGetDesc(op1->AsLclVar())->IsAddressExposed())
+    {
+        GenTree* addOp1 = op2->gtGetOp1();
+        GenTree* addOp2 = op2->gtGetOp2();
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            (addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum()) &&
+            ((op1->TypeIs(TYP_LONG) && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
+             comp->compOpportunisticallyDependsOn(InstructionSet_BMI1))
+            )
+        {
+            LIR::Use use;
+            if (BlockRange().TryGetUse(andNode, &use))
+            {
+                // If we allow the use of BSLR where the parent is NE or EQ it will generate "blsr, test"
+                // if we prevent the use of BSLR here then "OptimizeConstCompare" may generate TEST_NE or
+                //   TEST_EQ leading to a cheaper "lea, test"
+                GenTree* user = use.User();
+                if (!user->OperIs(GT_EQ, GT_NE))
+                {
+                    GenTreeHWIntrinsic* blsrNode =
+                        comp->gtNewScalarHWIntrinsicNode(andNode->TypeGet(), op1,
+                                                         andNode->TypeIs(TYP_INT)
+                                                             ? NamedIntrinsic::NI_BMI1_ResetLowestSetBit
+                                                             : NamedIntrinsic::NI_BMI1_X64_ResetLowestSetBit);
+
+                    JITDUMP("Lower: optimize AND(X, ADD(X, -1)): ");
+                    DISPNODE(andNode);
+                    JITDUMP("Replaced with: ");
+                    DISPNODE(blsrNode);
+
+                    use.ReplaceWith(blsrNode);
+
+                    BlockRange().InsertBefore(andNode, blsrNode);
+                    BlockRange().Remove(andNode);
+                    BlockRange().Remove(op2);
+                    BlockRange().Remove(addOp1);
+                    BlockRange().Remove(addOp2);
+                    JITDUMP("Remove [%06u], [%06u]\n", andNode->gtTreeID, andNode->gtTreeID);
```

Comment:
This `JITDUMP` is not terribly correct (prints the same node twice, and doesn't print the rest of the removed nodes). I think it could be removed, or if not, it should print IDs of all the removed nodes.

(It would be nice to print the "Removing unused node" message and each of them separately with `DISPNODE`, but it's not worth the code)
(We should probably add an option to `LIR::Range::Remove` to print this message, activated with a parameter, but that's unrelated cleanup)

Reply:
Removed. We Jitdump should provide useful information. Again this is a case where I looked at an existing function and did the same thing though so code quality elsewhere doesn't seem to meet the current standard.